upstairs/src/upstairs.rs

```diff
@@ -0,0 +1,3432 @@
+                let r = tx.send(out);
+                if r.is_err() {
+                    warn!(self.log, "control message reply failed");
+                }
+            }
+        }
+    }
+
+    /// Checks if a repair is possible. If so, checks if any Downstairs is in
+    /// the [DsState::LiveRepairReady] state, indicating it needs to be
+    /// repaired. If a Downstairs needs to be repaired, try to start repairing
+    /// it. When starting the repair fails, this function will schedule a task
+    /// to retry the repair by setting [Self::repair_check_interval].
+    ///
+    /// If this Upstairs is [UpstairsConfig::read_only], this function will move
+    /// any Downstairs from [DsState::LiveRepairReady] back to [DsState::Active]
+    /// without actually performing any repair.
+    pub(crate) async fn on_repair_check(&mut self) {
+        info!(self.log, "Checking if live repair is needed");
+        if !matches!(self.state, UpstairsState::Active) {
+            info!(self.log, "inactive, no live repair needed");
+            self.repair_check_interval = None;
+            return;
+        }
+
+        if self.cfg.read_only {
+            info!(self.log, "read-only, no live repair needed");
+            // Repair can't happen on a read-only downstairs, so short circuit
+            // here. There's no state drift to repair anyway, this read-only
+            // Upstairs wouldn't have caused any modifications.
+            for c in self.downstairs.clients.iter_mut() {
+                c.skip_live_repair(&self.state);
+            }
+            self.repair_check_interval = None;
+            return;
+        }
+
+        // Verify that all downstairs and the upstairs are in the proper state
+        // before we begin a live repair.
+        let repair_in_progress = self.downstairs.live_repair_in_progress();
+
+        let any_in_repair_ready = self
+            .downstairs
+            .clients
+            .iter()
+            .any(|c| c.state() == DsState::LiveRepairReady);
+
+        if repair_in_progress {
+            info!(self.log, "Live Repair already running");
+            // Queue up a later check if we need it
+            if any_in_repair_ready {
+                self.repair_check_interval = Some(deadline_secs(1.0));
+            } else {
+                self.repair_check_interval = None;
+            }
+        } else if !any_in_repair_ready {
+            self.repair_check_interval = None;
+            info!(self.log, "No Live Repair required at this time");
+        } else if !self.downstairs.start_live_repair(
+            &self.state,
+            self.guest.guest_work.lock().await.deref_mut(),
+            self.ddef.get_def().unwrap().extent_count().into(),
+            self.generation,
+        ) {
+            // It's hard to hit this condition; we need a Downstairs to be in
+            // LiveRepairReady, but for no other downstairs to be in Active.
```

Comment:
In the old world, this path existed to handle the case where the downstairs under repair hit an error and we
had to abort the repair.. but.. the downstairs that was being repaired that was kicked out actually came
back around fast enough that that downstairs now want to be repaired, but the upstairs has not yet finished
cleaning out the previous repair.  So, in essence, we were seeing our own repair still running.

In that case, we need to have this downstairs go off and wait a bit and let the aborted repair clean up and then try again.

Is that condition still possible in the one-task world?  I would think if we need IOs from the other downstairs to
ack back to the repair, it could still happen.

Reply:
Yeah, I think it still might be possible given the _exactly wrong_ sequence of events, so I think having the log + delay here is still valuable.

Reply:
I think your condition is possible but unlikely yes.

I mean more about the situation this path was originally for, the returning downstairs that was
under repair and needs to wait for the previous repair to finish before it can be repaired.

If that situation is still valid, then let's update the comment here to reflect it's a possible path
into this section.

Reply:
I think the case you're describing is now handled by this [earlier code](https://github.com/oxidecomputer/crucible/blob/9f5c855a4c9643391f59e43c8b874b052883326d/upstairs/src/upstairs.rs#L694-L708): if a live-repair was running but is now aborted, `Downstairs::live_repair_in_progress` will still return `true`, and we'll _also_ see `any_in_repair_ready = true`, so we'll set `repair_check_interval` at L708.

This is a change that went in recently; previously, `Downstairs::start_live_repair` was _also_ responsible for bailing out if a repair was already in progress, but it makes more sense to put that into `Downstairs::live_repair_in_progress`.

The conditional at L719-L723 is for the only remaining failure mode of `Downstairs::start_live_repair`, which is if we find no downstairs in `DsState::Active` to serve as sources for live-repair.

Reply:
Perfect, though we should bump the timer [in that earlier code](https://github.com/oxidecomputer/crucible/blob/9f5c855a4c9643391f59e43c8b874b052883326d/upstairs/src/upstairs.rs#L694-L708) to be 10 seconds.

Reply:
Oops, I thought I had already done that; fixed now!